### ISSUE TITLE
[0] Switching Between Wallets Shows Other Wallet's Accounts Until Reload

### ISSUE THREAD

**Issue by landry314** [bitshares/bitshares-ui#210]
In Settings -> Wallet when you change the active wallet then go to Settings -> Accounts, you see all the accounts from the last wallet still there.

You can switch tabs, make trades, do a bunch of things but until you RELOAD, they sit there.

The easy solution is to force a reload after switching wallets.

I am using the light wallet program on Mac OS X.

**Comment by wmbutler** [comment]
This is a react logic problem and definitely should be viewed as a high priority bug.

**Comment by wmbutler** [comment]
Related to #139

**Comment by calvinfroedge** [comment]
I can take this issue

**Comment by abitmore** [comment]
I think we can set a "default" account for each wallet when there are more than one accounts inside. When switching wallets, choose that one. Or use the first one. Or maybe use last clicked one.

**Comment by abitmore** [comment]
By the way I don't think this is very high priority. Most users have only one wallet in the browser.

**Comment by landry314** [comment]
Check out issue #248 I just created. I believe these issues can be solved with the same fix.

I agree, @abitmore, this is not HIGH priority because most users never create another wallet, but if they do, being confronted with bizarre behavior like this is very off-putting and does not instill trust. If we want bitshares to "replace the legacy financial system" then we must have a GUI which is seamless so that bitshares can be 100% trusted to store our money.

These bugs must be solved after the high priority bugs. :-)

**Comment by wmbutler** [comment]
The reason I view this as a high priority bug is that it would allow a user to perform operations on the wrong account. Even though it may affect few people, it's fairly serious in nature.

**Comment by calvinfroedge** [comment]
@wmbutler @landry314 Cannot reproduce. Please see attached screencast. Perhaps the confusion is that you actually need to click the "change to" button?

![issue-210](https://user-images.githubusercontent.com/632938/29436288-6d2af964-8370-11e7-9a4d-7e4d1b516d22.gif)

**Comment by wmbutler** [comment]
Holding until we get further clarification.

**Comment by landry314** [comment]
No, that is not it, @calvinfroedge. After doing what you show in your screencast THEN clicking on the "change to NAME" button, the wallet does not change, at least for me: it is still the old wallet until I RELOAD. Reloading is the only way I get the wallet to actually switch after changing it.

Can anyone try this and confirm the same result happens for them?

Oh, I am using the Mac OS light client, btw.

**Comment by calvinfroedge** [comment]
@landry314 Screencast would be helpful. If you can reproduce the issue when creating accounts from a new wallet, even better. I know it sucks, but the only way I can resolve is if I can reproduce.

Thanks for your time in reporting this issue.

**Comment by landry314** [comment]
Ah, wait a second, it might just be taking a long time. I just tried it switching from the default to the empty test wallet and after 2 minutes each on 2 tests nothing ever happened but switching from the empty test wallet to the default wallet, it took over 20 seconds one time and about 10 seconds the other time to switch over. That is very weird.

Also, a very weird thing happened in the meantime. On switching from the default wallet to the empty test wallet, I tried to hide an account before I had to force reload the new wallet. When I clicked "Hide" the account disappeared from the list. After reloading it appeared in the empty test wallet as the only account there. I switched back to the default wallet and that account was back in the default wallet, nothing wrong there, but upon switching to the (now not empty) test wallet I see the account is permanently now in that wallet.

This means I have found a inadvertent way to permanently duplicate an account from one wallet to another...

Also, when I first went back to the default wallet after hiding the account to see if it came back, it came back but it was no longer hidden AND it was placed in the top pulldown menu to choose active accounts. It was also now the active account. It is not starred nor in the Followed accounts so switching off of it in the top pulldown made it disappear from that menu.

This is very wonky. I hope it is safe though. I think I want to restore my wallet from backup because I don't know what else could be wrong.

**Comment by calvinfroedge** [comment]
@landry314 One thing at a time. Clear reproduction steps. Screencasts are best. Try the [licecap](https://www.cockos.com/licecap/) tool to create screencasts like the one I made above.

If there are issues let's work through them separately, one at a time.

**Comment by landry314** [comment]
It can't be separated because it is all tied together. "Hiding an account while in the middle of switching between two wallets before the reload happens causes the account to duplicate into both wallets and become active on the second wallet"? It is just silly.

The wallet and account system is so poorly written it can not be fixed by focusing on one "bug" at a time. It needs an epic.

I am sorry this is messy and hard to follow as described in words but the code is messy and I really don't know how to separate it without seeing the underlying structure. This is a job for a coder to get in there and really track the whole system to see what can be salvaged for a rewrite.

Check out my comment here: https://github.com/bitshares/bitshares-ui/issues/248#issuecomment-322556706 . In it I explain how five separate issues are tied together and how they can be solved together. If an epic was created called "Wallet and Account System Overhaul" then maybe all these things could be studied together. I don't think this can be solved in simply a bug squashing mission, but it would be nice if I were wrong.

Now it is 5 issues that I feel are tied together: #196, #207, #210, #215, and #248. I have explained each one the best I can in every one of these issue pages. I know it is a bit messy, I can't really add any more without going into the code myself. I have tried my best to explain each problem.

I have studied some coding but it was Pascal, C++, Ada95, CSS, nothing that applies here. I would need to learn node js, right? I am interested to learn but I am going off the grid now for a week into nature for the solar eclipse... See you on the other side. I hope you can find out what is going on with this code! Good luck!

**Comment by landry314** [comment]
Oh, and you might want to even throw in issue #216 though it barely matters.

**Comment by landry314** [comment]
And I am using the latest build on bitshares.org/download, which is BitShares 2 (March 28th, 2016 / GUI May 22nd, 2017). I know there is a newer one but I haven't tried to build it myself. Issue #224 would solve this! haha.

**Comment by wmbutler** [comment]
This appears to be related to latency to me. It's taking your client so long to talk to the node server that your client is attempting to perform actions prior to fully getting new account data. Please state how may wallets and how many accounts you have and your latency to your node server.

**Comment by landry314** [comment]
2 wallets but only because I created one to test with.
27 accounts in my one main wallet.
416ms latency at the moment connected to Dallas dexnode.net but i've used different nodes depending on which is better.

**Comment by wmbutler** [comment]
Since neither @calvinfroedge nor I can reproduce this, I think we are going to need to see a screenshare. You may want to wait until after the 9/1 release date to see if we solved any of these issues. I know for a fact that we updated a few libraries that were causing excess traffic.

**Comment by landry314** [comment]
Yes, I will wait until after September 1 and see if it is fixed. Will the download be available on bitshares.org/download?

It is still happening for me, I just checked. And that account is definitely permanently duplicated to both wallets now.

**Comment by wmbutler** [comment]
Yes, the download for all light wallets will be released then. Regarding your second comment, yeah, since nothing has changed, it makes sense that you would witness the same behavior.

**Comment by wmbutler** [comment]
@landry314 please let me know if this still happens for you. If not, please close. Be sure to DL the latest client.

**Comment by landry314** [comment]
Yes, it still happens exactly the same when switching from my wallet with many accounts to my wallet with only one account, it partially switches but still shows all the accounts in the Settings -> Accounts section until I reload (or maybe it changes after 2 minutes, I haven't waited).

I am using the Aug 30 build of the light client for Mac OS.

**Comment by landry314** [comment]
Wouldn't an easy fix just be to add a reload at the end of the function?

**Comment by wmbutler** [comment]
Please post a screenshare, ideally with your console visible. Thanks!

https://www.cockos.com/licecap/
